gravitee-node-api/src/main/java/io/gravitee/node/api/secrets/runtime/spec/Spec.java

```diff
@@ -0,0 +1,205 @@
+ *
+ * @param id              generated UUID. Optional when <code>isGenerated</code> is true
+ * @param name            unique per environment name when set.
+ *                        Constraints are:
+ *                        length is min 3,
+ *                        cannot start and end with non-alphanumeric chars,
+ *                        can contain '_', ' ', '.', '-' and  alphanumeric chars
+ * @param uri             mandatory string that designate the secret to fetch as specified by the secret manager.
+ *                        It follows the {@link SecretURL} syntax: /&lt;provider id or plugin id>/&lt;secret location>
+ * @param key             the key in the secret map (if known)
+ * @param usesDynamicKey  when the key is computed at runtime hence true when <code>key</code> param is null
+ * @param isGenerated     if the spec was generated from a reference, as opposed create by a user
+ * @param resolution      how the secret is resolved
+ * @param onErrorStrategy
+ * @param acls            access control list: where in definitions secrets are allowed to be resolved
+ * @param envId           mandatory environment ID in which this definition applies
+ * @author Benoit BORDIGONI (benoit.bordigoni at graviteesource.com)
+ * @author GraviteeSource Team
+ * @see Resolution
+ * @see ACLs
+ */
+public record Spec(
+    String id,
+    String name,
+    String uri,
+    String key,
+    boolean usesDynamicKey,
+    boolean isGenerated,
+    Resolution resolution,
+    Spec.OnErrorStrategy onErrorStrategy,
+    ACLs acls,
+    String envId
+) {
+    private static final char[] ALLOWED_NON_ALPHA_NUM_CHARS = new char[] { '_', ' ', '.', '-' };
+    static {
+        Arrays.sort(ALLOWED_NON_ALPHA_NUM_CHARS);
+    }
+
+    public Spec {
+        boolean ok = isGenerated ^ isNotBlank(id);
+        if (!ok) {
+            throw new IllegalArgumentException("spec is either generated or contains an id");
+        }
+        ok = usesDynamicKey ^ isNotBlank(key);
+        if (!ok) {
+            throw new IllegalArgumentException("spec either uses dynamic key of specifies one");
+        }
+        ok = isNotBlank(envId);
+        if (!ok) {
+            throw new IllegalArgumentException("spec must contain an envId");
+        }
+        ok = isNotBlank(uri);
+        if (!ok) {
+            throw new IllegalArgumentException("spec must contain a uri");
+        }
+        ok = uri.charAt(0) == SecretURL.URL_SEPARATOR;
+        if (!ok) {
+            throw new IllegalArgumentException("uri must start with '%s'".formatted(SecretURL.URL_SEPARATOR));
+        }
+        if (name != null) {
+            assertName(name);
+        }
+    }
+
+    /**
+     * Return uri and key concatenated as specified in {@link Spec#formatUriAndKey(String, String)}
+     * @return a string concat of uri and key
+     */
+    public String uriAndKey() {
+        return formatUriAndKey(uri, key);
```

Comment:
I don't think you need this redirection, it could be easier to understand if the code was mode here directly.

Reply:
The static method is used in plugin and here.